### PR TITLE
Replace K8S task watcher to K8S task informer

### PR DESCRIPTION
**Pull request by zhaohehuhu**
## Purpose of the pull request
I'd like to replace K8S task watcher to informer. Sometimes watcher may not receive full watcher events for a Job, which may make K8S task stuck in the wrong status(See [issue](https://github.com/fabric8io/kubernetes-client/issues/2323)) Besides, compared with watcher, informer can get all events that job had even before informer was running.  Informer provides a better way to cache and retrieve resources from the Kubernetes API server efficiently based on watcher, which can highly reduce API server workload.

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

**Comment by Radeity** [comment]
Hi @zhaohehuhu, you should better link an issue.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java [hunk 1]
```diff
@@ -193,7 +195,55 @@ public void buildK8sJob(K8sTaskMainParameters k8STaskMainParameters) {
...
+                } finally {
+                    LogUtils.removeTaskInstanceLogFullPathMDC();
+                }
 
+            }
+            @Override
+            public void onDelete(Job job, boolean b) {
+                log.info("[K8sJobExecutor-{}] job got deleted", job.getMetadata().getName());
+
+            }
```
Comment (Radeity) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java:232]: Have to add handler logic for delete for delete event.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java [hunk 2]
```diff
@@ -193,7 +195,55 @@ public void buildK8sJob(K8sTaskMainParameters k8STaskMainParameters) {
...
     }
+    public void registerBatchJobFormer(Job job, String taskInstanceId, TaskResponse taskResponse) {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        ResourceEventHandler resourceEventHandler =  new ResourceEventHandler<Job>() {
+            @Override
+            public void onAdd(Job job) {
+                log.info("[K8sJobExecutor-{}] job got added", job.getMetadata().getName());
+            }
+            @Override
+            public void onUpdate(Job job, Job t1) {
```
Comment (Radeity) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java:212]: Please add blank line before each method.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java [hunk 3]
```diff
@@ -193,7 +195,55 @@ public void buildK8sJob(K8sTaskMainParameters k8STaskMainParameters) {
...
+        } catch (InterruptedException e) {
+            log.error("job failed in k8s: {}", e.getMessage(), e);
+            Thread.currentThread().interrupt();
+            taskResponse.setExitStatusCode(EXIT_CODE_FAILURE);
+        } catch (Exception e) {
+            log.error("job failed in k8s: {}", e.getMessage(), e);
+            taskResponse.setExitStatusCode(EXIT_CODE_FAILURE);
+        }
+    }
+    @Deprecated
```
Comment (Radeity) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java:254]: Ditto.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java [hunk 4]
```diff
@@ -193,7 +195,55 @@ public void buildK8sJob(K8sTaskMainParameters k8STaskMainParameters) {
...
+                try {
+                    LogUtils.setTaskInstanceLogFullPathMDC(taskRequest.getLogPath());
+                    log.info("[K8sJobExecutor-{}] job got updated", job.getMetadata().getName());
+                    int jobStatus = getK8sJobStatus(job);
+                    log.info("job {} status {}", job.getMetadata().getName(), jobStatus);
+                    if (jobStatus == TaskConstants.RUNNING_CODE) {
+                        return;
+                    }
+                        setTaskStatus(jobStatus, taskInstanceId, taskResponse);
+                        countDownLatch.countDown();
```
Comment (Radeity) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java]: Please align together with last line.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java [hunk 5]
```diff
@@ -193,7 +195,55 @@ public void buildK8sJob(K8sTaskMainParameters k8STaskMainParameters) {
...
+            public void onAdd(Job job) {
+                log.info("[K8sJobExecutor-{}] job got added", job.getMetadata().getName());
+            }
+            @Override
+            public void onUpdate(Job job, Job t1) {
+                try {
+                    LogUtils.setTaskInstanceLogFullPathMDC(taskRequest.getLogPath());
+                    log.info("[K8sJobExecutor-{}] job got updated", job.getMetadata().getName());
+                    int jobStatus = getK8sJobStatus(job);
+                    log.info("job {} status {}", job.getMetadata().getName(), jobStatus);
```
Comment (Radeity) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java]: Can we print the log only once?

**Review by Radeity — COMMENTED** [review]
BTW, do we really have to use informer for `Job`?

**Comment by Radeity** [comment]
@Gallardot  PTAL

**Comment by ruanwenjun** [comment]
Is there any way to add K8sTask test case?

**Comment by Gallardot** [comment]
> Is there any way to add K8sTask test case?

+1. Adding test cases for K8sTask is highly necessary. 
 
Additionally, I am not very familiar with the difference between watcher and informer in the k8s java-client.  Can you explain the difference between watcher and informer, or provide some reference documents? @zhaohehuhu

**Comment by zhaohehuhu** [comment]
> Is there any way to add K8sTask test case?

There is already a UT for K8sTask. But it's not easy to do a full test case that covers every things in K8S task.

**Comment by zhaohehuhu** [comment]
> > Is there any way to add K8sTask test case?
> 
> +1. Adding test cases for K8sTask is highly necessary.
> 
> Additionally, I am not very familiar with the difference between watcher and informer in the k8s java-client. Can you explain the difference between watcher and informer, or provide some reference documents? @zhaohehuhu

I can provide the doc. But you can click on the issue I mentioned earlier for details now.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java (hunk 1 above)
Comment (zhaohehuhu) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java:232]: For delete event, I believe it should be enough now.  no more work to  do.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java (hunk 2 above)
Comment (zhaohehuhu) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java:212]: done. Thanks.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java (hunk 3 above)
Comment (zhaohehuhu) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java:254]: done

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java (hunk 4 above)
Comment (zhaohehuhu) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java]: done

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java (hunk 5 above)
Comment (zhaohehuhu) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/k8s/impl/K8sTaskExecutor.java]: Sure.

**Pull request closed without merging**